### PR TITLE
[Merged by Bors] - Update wgpu to 0.12 and naga to 0.8

### LINKED ISSUE #3208: Pipelined rendering doesn't work on DX12
## Bevy version

main branch, e8412df0211f5684aa03459e1f47a5c40035f6c1

## Operating system & version

Windows 11

## What you did

Run any of the pipelined examples with WGPU_BACKEND="dx12" env variable. 

(For those who don't know, in powershell you can enter `$env:WGPU_BACKEND="dx12"` (including the $) to change the env variable for that terminal session).

## What you expected to happen

I expect them to crash, as this is an upstream issue in wgpu.

## What actually happened

They crashed.

## Additional information

Upstream issue: https://github.com/gfx-rs/wgpu/issues/2185

This is due to bevy enabling the TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES wgpu feature, which is implemented for DX12 but they forgot to expose it as a DX12 feature.

I tried adding the feature to the DX12 feature list and testing it with https://github.com/bevyengine/bevy/pull/3041's game_of_life_pipelined example (it uses a ReadWrite StorageTexture), but it still crashes.

My wgpu 0.11 branch that enables the feature: https://github.com/Elabajaba/wgpu/tree/dx12-feature-0.11

### LINKED ISSUE #3352: wgpu::backend::direct: Error in Adapter::request_device: unsupported features were requested: TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES
## Bevy version0.5.0

The release number or commit hash of the version you're using.

## Operating system & version

 Windows 10,
Error report
D:\Program Files\rust\bevy2021>cargo run --example 3d_scene
   Compiling bevy v0.5.0 (D:\Program Files\rust\bevy2021)
    Finished dev [unoptimized + debuginfo] target(s) in 55.94s
     Running `target\debug\examples\3d_scene.exe`
2021-12-16T09:12:11.025569Z  INFO bevy_render::renderer: AdapterInfo { name: "AMD Radeon(TM) Vega 8 Graphics", vendor: 4098, device: 5597, device_type: DiscreteGpu, backend: Dx12 }
2021-12-16T09:12:11.026104Z ERROR wgpu::backend::direct: Error in Adapter::request_device: unsupported features were requested: TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: RequestDeviceError', crates\bevy_render\src\renderer\mod.rs:91:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: process didn't exit successfully: `target\debug\examples\3d_scene.exe` (exit code: 101)

I don't know how to fix it. The latest version was downloaded as soon as it was updated today.
QQ拼音截图未命名

### PR DESCRIPTION
**Pull request by vabka**
# Objective

Fixes #3352
Fixes #3208

## Solution

- Update wgpu to 0.12
- Update naga to 0.8
- Resolve compilation errors
- Remove [[block]] from WGSL shaders (because it is depracated and now wgpu cant parse it)
- Replace `elseif` with `else if` in pbr.wgsl

### CODE REVIEW

File: assets/shaders/custom_material.wgsl [hunk 1]
```diff
@@ -1,4 +1,4 @@
-[[block]]
+
```
Comment (bjorn3) [inline assets/shaders/custom_material.wgsl:1]:
```suggestion
```

File: crates/bevy_pbr/src/render/depth.wgsl [hunk 2]
```diff
@@ -1,7 +1,7 @@
 #import bevy_pbr::mesh_struct
 
 // NOTE: Keep in sync with pbr.wgsl
-[[block]]
+
```
Comment (bjorn3) [inline crates/bevy_pbr/src/render/depth.wgsl:4]:
```suggestion
```
same everywhere else

File: crates/bevy_gltf/Cargo.toml [hunk 3]
```diff
@@ -29,4 +29,4 @@ thiserror = "1.0"
 anyhow = "1.0.4"
 base64 = "0.13.0"
 percent-encoding = "2.1"
-wgpu = "0.11.0"
+wgpu = { version = "0.12.0" }
```
Comment (bjorn3) [inline crates/bevy_gltf/Cargo.toml:32]:
```suggestion
wgpu = "0.12.0"
```
looks a bit nicer and is more consistent.

File: crates/bevy_pbr/src/render/mesh.rs [hunk 4]
```diff
@@ -210,10 +207,7 @@ impl FromWorld for MeshPipeline {
                 BindGroupLayoutEntry {
                     binding: 3,
                     visibility: ShaderStages::FRAGMENT,
-                    ty: BindingType::Sampler {
-                        comparison: true,
-                        filtering: true,
-                    },
+                    ty: BindingType::Sampler(SamplerBindingType::Comparison),
```
Comment (bjorn3) [inline crates/bevy_pbr/src/render/mesh.rs:210]: Can someone verify if this isn't a behavorial change.

File: crates/bevy_gltf/Cargo.toml (hunk 3 above)
Comment (cart) [inline crates/bevy_gltf/Cargo.toml:32]: when theres no other configuration we favor the simpler `wgpu = "0.12.0"` format

File: crates/bevy_sprite/Cargo.toml [hunk 5]
```diff
@@ -31,3 +31,4 @@ guillotiere = "0.6.0"
 thiserror = "1.0"
 rectangle-pack = "0.4"
 serde = { version = "1", features = ["derive"] }
+wgpu = { version = "0.12.0" }
```
Comment (bjorn3) [inline crates/bevy_sprite/Cargo.toml:34]:
```suggestion
wgpu = "0.12.0"
```

also nit: missing trialing newline

File: crates/bevy_pbr/src/render/depth.wgsl (hunk 2 above)
Comment (vabka) [reply crates/bevy_pbr/src/render/depth.wgsl]: fixed

File: crates/bevy_sprite/src/render/mod.rs [hunk 6]
```diff
@@ -351,15 +349,15 @@ pub fn prepare_sprites(
                         handle: current_batch_handle.clone_weak(),
                         z: last_z,
                         colored: true,
-                    },));
+                    }, ));
```
Comment (bjorn3) [inline crates/bevy_sprite/src/render/mod.rs]: Doesn't rustfmt fail with this change?

File: crates/bevy_ui/Cargo.toml [hunk 7]
```diff
@@ -29,7 +29,8 @@ bevy_utils = { path = "../bevy_utils", version = "0.5.0" }
...
 # other
 stretch = "0.3.2"
-serde = {version = "1", features = ["derive"]}
+serde = { version = "1", features = ["derive"] }
 smallvec = { version = "1.6", features = ["union", "const_generics"] }
 bytemuck = { version = "1.5", features = ["derive"] }
-crevice = { path = "../crevice", version = "0.8.0", features = ["glam"] }
\ No newline at end of file
+crevice = { path = "../crevice", version = "0.8.0", features = ["glam"] }
+wgpu = { version = "0.12.0" }
```
Comment (bjorn3) [inline crates/bevy_ui/Cargo.toml:36]: same re version specification and the missing trailing newline nit.

File: crates/bevy_gltf/Cargo.toml (hunk 3 above)
Comment (vabka) [reply crates/bevy_gltf/Cargo.toml]: fixed

**Comment by bjorn3** [comment]
Thanks for the PR! I have a couple of nits and one thing someone else needs to check, but apart from that I don't see any issues.

File: crates/bevy_sprite/Cargo.toml (hunk 5 above)
Comment (vabka) [reply crates/bevy_sprite/Cargo.toml]: fixed

File: crates/bevy_pbr/src/render/mesh_view_bind_group.wgsl [hunk 8]
```diff
@@ -35,7 +34,7 @@ struct DirectionalLight {
 
 let DIRECTIONAL_LIGHT_FLAGS_SHADOWS_ENABLED_BIT: u32 = 1u;
 
-[[block]]
+
```
Comment (mrk-its) [inline crates/bevy_pbr/src/render/mesh_view_bind_group.wgsl:37]: let's remove these empty lines

File: crates/bevy_ui/Cargo.toml (hunk 7 above)
Comment (vabka) [reply crates/bevy_ui/Cargo.toml]: fixed

File: crates/bevy_pbr/src/render/mesh_view_bind_group.wgsl (hunk 8 above)
Comment (vabka) [reply crates/bevy_pbr/src/render/mesh_view_bind_group.wgsl]: Hah. Missed this one :)
Fixed

File: crates/bevy_gltf/Cargo.toml (hunk 3 above)
Comment (vabka) [reply crates/bevy_gltf/Cargo.toml]: fixed

File: assets/shaders/custom_material.wgsl (hunk 1 above)
Comment (vabka) [reply assets/shaders/custom_material.wgsl]: fixed

File: crates/bevy_pbr/src/render/mesh.rs (hunk 4 above)
Comment (vabka) [reply crates/bevy_pbr/src/render/mesh.rs:213]: +1
Also not sure

File: crates/bevy_sprite/src/render/mod.rs (hunk 6 above)
Comment (vabka) [reply crates/bevy_sprite/src/render/mod.rs]: I suppose it is the intellij-rust :)

File: crates/bevy_pbr/src/render/mesh_view_bind_group.wgsl [hunk 9]
```diff
@@ -56,18 +54,18 @@ struct Lights {
     n_directional_lights: u32;
 };
 
-[[block]]
+
```
Comment (bjorn3) [inline crates/bevy_pbr/src/render/mesh_view_bind_group.wgsl:57]:
```suggestion
```
missed a couple in this file.

File: crates/bevy_render/src/render_resource/shader.rs [hunk 10]
```diff
@@ -849,7 +842,7 @@ void bar() { }
     fn process_nested_shader_def_outer_defined_inner_not() {
         #[rustfmt::skip]
     const EXPECTED: &str = r"
-[[block]]
+
```
Comment (bjorn3) [inline crates/bevy_render/src/render_resource/shader.rs:845]: And in this file.

**Review by bjorn3 — COMMENTED** [review]
Looks like two files still have extra newlines. The rest has been fixed.

**Comment by vabka** [comment]
finally removed all blank lines

File: crates/bevy_pbr/src/render/mesh.rs (hunk 4 above)
Comment (vabka) [reply crates/bevy_pbr/src/render/mesh.rs:213]: I dont know what these options should affect. But with other SamplingBindingType::Filtering in panics.
Result looks pretty same.

**Comment by mockersf** [comment]
Could you also remove this line?
https://github.com/bevyengine/bevy/blob/8c250919e3f8ae9d485150966c4c8b7f38d8e543/deny.toml#L61

It's a duplicate dependency that's no longer duplicated with the wgpu update 👍

**Comment by vabka** [comment]
@mockersf Why it was added? What 'raw-window-handle' does?
(Sorry if it is a stupid question. I am a newbie in rust)

**Comment by mockersf** [comment]
Not a stupid question, no worries!

Bevy is using [cargo-deny](https://github.com/EmbarkStudios/cargo-deny) to help control the dependency list. With it, we can make CI fail for things like unmaintained dependencies, or duplicated dependencies.

In this case, `raw-window-handle` is a dependency added by both `wgpu` and `winit`, and it was updated in Bevy with the last `winit` update, while still depending on the previous version through `wgpu`. Now that `wgpu` is also updated, the last version of `raw-window-handle` is used everywhere so we don't need to accept it as a duplicated dependency.

[`raw-window-handle`](https://github.com/rust-windowing/raw-window-handle) is... I don't know exactly, but it seems it's used to handle windows on all OS

**Review by mockersf — APPROVED** [review]
works on macOS

**Comment by vabka** [comment]
Thank you for such explanation :)

**Comment by Josh015** [comment]
@vabka @mockersf `raw-window-handle` just lets a user pass a window handle to a graphics API to render into in a platform-agnostic way. It moves that responsibility out of their code and into the crates they depend upon.

**Comment by vabka** [comment]
> @vabka @mockersf `raw-window-handle` just lets a user pass a window handle to a graphics API so it knows where to render. Specifically, it hides the platform-specific details of this operation and moves that responsibility out of the user's code and into the crates they depend upon.

I mean "what bad thing it does so it was banned". I know what is window handle.
But thanks :)

File: crates/bevy_pbr/src/render/mesh.rs (hunk 4 above)
Comment (cart) [reply crates/bevy_pbr/src/render/mesh.rs:213]: Yeah I think comparison is fine for now. We might be able to rein this in, but this seems like the right choice for the current impl.

**Comment by cart** [comment]
It wasn't banned, the "failure on duplicate dependency" was just suppressed.

**Pull request closed without merging**